GitUI/Avatars/AvatarService.cs

```diff
@@ -1,12 +1,126 @@
-﻿using GitUI.Properties;
+﻿using System;
+using System.Linq;
+using GitCommands;
+using GitUI.Properties;
 
 namespace GitUI.Avatars
 {
     public static class AvatarService
     {
-        private static readonly InitialsAvatarGenerator AvatarGenerator = new InitialsAvatarGenerator();
-        public static IAvatarProvider Default { get; }
-            = new BackupAvatarProvider(new AvatarMemoryCache(new AvatarPersistentCache(new AvatarDownloader(AvatarGenerator), AvatarGenerator)),
-                Images.User80);
+        private static readonly InitialsAvatarProvider InitialsAvatarProvider;
+        private static readonly StaticImageAvatarProvider UserImageAvatarProvider;
+
+        private static readonly HotSwapAvatarProvider HotSwapProvider;
```

Comment:
I think we don't needed HotSwapAvatarProvider class. We catch the exception higher in the call hierarchy.

Reply:
A lot of other gitextension Code keeps a reference to the provider. Without the ability to hotswap the inner provider those components would have references to an outdated avatar provider.

Reply:
What code are you talking about? This is a private field.

Reply:
The HotSwapProvider is the inner avatar provider (before caching) and it's indirectly exposed by the `DefaultProvider` Property.

There a hand full of classes keep a reference to that `DefaultProvider`, so if we later update the DefaultProvider, because the settings changed, and we just throw the old one out and set a new one, all classes that still have a reference to the old provider would query avatars with an outdated provider / outdated settings.

This is why the HotSwapProvider exists, so the effective inner provider can be changed without changing the Reference to the DefaultProvider, and without rebuilding the cache instances.

Alternatively we could rewrite everything else to always use the static property (but this would tightly couple a lot of classes to the AvatarService Singleton - which could make testing way harder).

Another way would be to provide some sort of update mechanism/event, something like Observables, but that imply even more big design changes in gitextensions.

Reply:
Move state to AvatarService. It main facade.

Can be deleted HotSwapAvatarProvider.
Can be deleted IAvatarCacheCleaner.
Can be simplified FileSystemAvatarProvider (no state).
Can be simplified MemoryAvatarProvider (no state).
Can be deleted public static IAvatarProvider DefaultProvider { get; }
Can be deleted public static IAvatarCacheCleaner CacheCleaner { get; }

Look at GetAvatarAsync how much complex logic was hidden.

If needed not use AvatarService as static class but cache we still need as static.

There are many things in the complex. And you say here a crutch because there is a crutch and so on.

```
private const int MaxCachedImages = 30;

private readonly _fileSystem = new FileSystem();
private readonly _cache = new MruCache<(string email, int imageSize), Image>(MaxCachedImages);

public static async Task<Image> GetAvatarAsync(string email, string name, int imageSize)
{
    var fileSystemAvatarProvider = new FileSystemAvatarProvider(_fileSystem);
    var memoryAvatarProvider = new MemoryAvatarProvider(_cache);

    var image = await memoryAvatarProvider
        .GetAvatarAsync(email, name, imageSize);

    if (image is null)
    {
        image = await fileSystemAvatarProvider
            .GetAvatarAsync(email, name, imageSize);

        if (image is null)
        {
            image = await HotSwapProvider // should removed
                .GetAvatarAsync(email, name, imageSize);

            if (image is null)
            {
                image = await UserImageAvatarProvider
                    .GetAvatarAsync(email, name, imageSize);
            }
            else
            {
                CacheToFile(email, imageSize, image);
            }
        }
        else
        {
            CacheToMemory(email, imageSize, image);
        }
    }

    return image;
}

private static void CacheToFile(string email, int imageSize, Image image)
{
    var cacheDir = AppSettings.AvatarImageCachePath;
    var path = Path.Combine(cacheDir, $"{email}.{imageSize}px.png");

    if (!_fileSystem.Directory.Exists(cacheDir))
    {
        _fileSystem.Directory.CreateDirectory(cacheDir);
    }

    try
    {
        using var output = _fileSystem.File
            .OpenWrite(path);

        image.Save(output, ImageFormat.Png);
    }
    catch
    {
        // do nothing
    }
}

private static void CacheToMemory(string email, int imageSize, Image image)
{
    lock (_cache)
    {
        cache.Add((email, imageSize), image);
    }
}

private async Task ClearCacheAsync()
{
    lock (_cache)
    {
        _cache.Clear();
    }
    
    var cachePath = AppSettings.AvatarImageCachePath;

    if (_fileSystem.Directory.Exists(cachePath))
    {
        await Task.Run(() =>
        {
            foreach (var file in _fileSystem.Directory.GetFiles(cachePath))
            {
                try
                {
                    _fileSystem.File.Delete(file);
                }
                catch
                {
                    // do nothing
                }
            }
        });
    }
}
```